tiledb/sm/filesystem/vfs.cc

```diff
@@ -51,32 +51,92 @@
 using namespace tiledb::common;
 using tiledb::common::filesystem::directory_entry;
 
-namespace tiledb {
-namespace sm {
+namespace tiledb::sm {
 
 /* ********************************* */
 /*     CONSTRUCTORS & DESTRUCTORS    */
 /* ********************************* */
 
-VFS::VFS()
-    : stats_(nullptr)
-    , init_(false)
-    , read_ahead_size_(0)
-    , compute_tp_(nullptr)
-    , io_tp_(nullptr) {
+VFS::VFS(
+    stats::Stats* const parent_stats,
+    ThreadPool* const compute_tp,
+    ThreadPool* const io_tp,
+    const Config& config)
+    : stats_(parent_stats->create_child("VFS"))
+    , config_(config)
+    , compute_tp_(compute_tp)
+    , io_tp_(io_tp)
+    , vfs_params_(VFSParameters(config)) {
+  Status st;
+  assert(compute_tp);
+  assert(io_tp);
+
+  // Construct the read-ahead cache.
+  read_ahead_cache_ = tdb_unique_ptr<ReadAheadCache>(
+      tdb_new(ReadAheadCache, vfs_params_.read_ahead_cache_size_));
+
+#ifdef HAVE_HDFS
+  supported_fs_.insert(Filesystem::HDFS);
+  hdfs_ = tdb_unique_ptr<hdfs::HDFS>(tdb_new(hdfs::HDFS));
+  st = hdfs_->init(config_);
+  if (!st.ok()) {
+    throw std::runtime_error("[VFS::VFS] Failed to initialize HDFS backend.");
+  }
+#endif
+
+#ifdef HAVE_S3
+  supported_fs_.insert(Filesystem::S3);
+  st = s3_.init(stats_, config_, io_tp_);
+  if (!st.ok()) {
+    throw std::runtime_error("[VFS::VFS] Failed to initialize S3 backend.");
+  }
+#endif
+
 #ifdef HAVE_AZURE
   supported_fs_.insert(Filesystem::AZURE);
+  st = azure_.init(config_, io_tp_);
+  if (!st.ok()) {
+    throw std::runtime_error("[VFS::VFS] Failed to initialize Azure backend.");
+  }
 #endif
+
 #ifdef HAVE_GCS
   supported_fs_.insert(Filesystem::GCS);
+  st = gcs_.init(config_, io_tp_);
+  if (!st.ok()) {
+    // We should print some warning here, LOG_STATUS only prints in
+    // verbose mode. Since this is called in the init of the context, we
+    // can't return the error through the normal set it on the context.
+    throw StatusException(Status_GCSError(
+        "GCS failed to initialize, GCS support will not be available in this "
+        "context: " +
+        st.message()));
+  }
 #endif
-#ifdef HAVE_HDFS
-  supported_fs_.insert(Filesystem::HDFS);
+
+#ifdef _WIN32
+  win_.init(config_, io_tp_);
+#else
+  posix_.init(config_, io_tp_);
 #endif
+
+  supported_fs_.insert(Filesystem::MEMFS);
+}
+
+VFS::~VFS() {
```

Comment:
Couldn't all this code go into the S3 class destructor instead? There doesn't seem to be an vfs-wide reason to do it here.